### PR TITLE
second round of refactorings for universes

### PR DESCRIPTION
**Pull request by nikomatsakis**
A second round of (what I believe to be) "no functional change" refactorings, taken from my universes branch.

r? @scalexm

### CODE REVIEW

File: src/librustc/infer/higher_ranked/mod.rs [hunk 1]
```diff
@@ -593,7 +593,7 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
     where
         T : TypeFoldable<'tcx>,
     {
-        let new_universe = self.create_subuniverse();
+        let new_universe = self.create_superuniverse();
```
Comment (arielb1) [inline src/librustc/infer/higher_ranked/mod.rs]: I'm not sure I prefer "superuniverse" to "subuniverse". Universes naturally form a tree, and the "subelement" relationship goes from a universe to its sub/superuniverse.

**Comment by nikomatsakis** [comment]
@arielb1 

> I'm not sure I prefer "superuniverse" to "subuniverse". Universes naturally form a tree, and the "subelement" relationship goes from a universe to its sub/superuniverse.

It seems like the use of "subuniverse" conflicts with methods like `is_subset_of` -- i.e., U0 is a "subset" of U1. Perhaps we can find some other terminology. I've found it pretty important when writing this code to have something that gives the right intuition.

Maybe "extension" and "extended" is a better terminology?

e.g., 

- `create_subuniverse` => `create_extended_universe`
- `is_subset_of` => `is_extended_by`

and so forth?

(For that matter, `create_superuniverse` should perhaps just be called `create_universe` or `create_fresh_universe`, with the "extension" aspect of it left to doc comments -- i.e., "creates a fresh universe that extends all previous universes")

**Comment by nikomatsakis** [comment]
Rebased and tidied. Went and tried the renaming I proposed. I wound up at a slightly different place, as you can see in the comment from the final commit:

---

remove the sub/super terminology for universes

Instead, we talk about:

- creating the "next" universe
- universes "extending" one another
- and `u1.can_name(u2)`, meaning that `u1` contains all names from `u2`

**Review by scalexm — APPROVED** [review]

**Comment by Manishearth** [comment]
seems to be broken?

```
error: internal compiler error: librustc\infer\canonical\canonicalizer.rs:175: unexpected region in query response: `ReEarlyBound(0, 'i)`
```

in the rollup

https://ci.appveyor.com/project/rust-lang/rust/builds/19415197/job/kex1uawo1pxeg0ml

**Comment by nikomatsakis** [comment]
Whoops, I thought I had patched that problem.

**Pull request merged**